### PR TITLE
fix inline code blocks for native platforms

### LINKED ISSUE #986: Blocks of `quoted` text render incorrectly on mobile
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!
___

View the job on Upwork [here](https://www.upwork.com/jobs/~01fc428a62f97f17a2).

**Platform - version:**  

iOS and Android, build 263

**Action Performed (reproducible steps):**

1. Type a sentence with code block formatting (single ticks) in the middle, like this:

This is a test to `see how code renders in the middle` of a sentence on mobile.

2. Notice that it renders strangely by either breaking the line or overflowing.

**Expected Result:**

It should render the same way it does here on GitHub.

This is a test to `see how code renders in the middle` of a sentence on mobile.

**Actual Result:**

The line either breaks or overflows where the code formatting starts/ends

**Notes/Photos/Videos:** 

We've seen multiple results from this bug. Here are some examples:

101217023-6e804a00-3635-11eb-9bb9-2e37a0eeccbb

102281699-4c22f080-3f94-11eb-878b-345b9d978e46

### PR DESCRIPTION
**Pull request by parasharrajat**
Please review @sketchydroide.

#### Problem

We (Including other contributors) tried using a combination of Text & View tag to achieve the above UI but there are a couple of issues as follows:
 1. Text is rendered as a block box but we want inline text.
 2. In Another way, we can get the inline text but border and border-radius are not being rendered.

#### How I fixed it?
I had an idea of tokenizing the text. Then apply the style selectively. This is working well. I have made a small snap here.

1. we create one wrapper component which divides the text into tokens(words) divided based on word break. This component put the text tokens inline into a view and renders it.
2. We can then pass styles selectively to tokens and manage the border-radius for the first and last tokens.

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes #986

### Tests / QA Steps

- Type a sentence with code block formatting (single ticks) in the middle, like this:
   This is a test to see `how code renders in the middle` of a sentence on mobile.
- Notice that it should render inline like normal text.

### Issue(Not blocker)
1. ~Inline Rendered code block is not vertically center with respect to the text.~ :heavy_check_mark: *Fixed*

### Tested On

- [x] iOS
- [x] Android
- [x] Web
- [x] Desktop

#### Web / Desktop

localhost_8080_r_68438375

#### IOS
I don't have an IOS device thus I mimicked the logic [here](https://snack.expo.io/@rajat_parashar/text-wrapping) and which can be seen working well.

image

### CODE REVIEW

**Comment by parasharrajat** [comment]
@shawnborton I hope this style matches the desired one.

**Comment by Luke9389** [comment]
Hey @parasharrajat, I'm noticing that this gets a bit clunky when there are longer words in the text. Here's what I'm talkin' about: 
Screen Shot 2021-04-08 at 3 44 23 PM

Think there's any way for us to fill in the white space with the code text style?

**Comment by shawnborton** [comment]
So if all of our message text is at 15px text size but with 20px line height, that means a normal looking piece of text should be 20px tall. I wonder if we could make our inline code (and code blocks for that matter) use a text size of 13px and a line height of 18px? This way the text feels more balanced, and when code wraps two lines, we get a small gap (2px) between the lines:
image

**Comment by parasharrajat** [comment]
@Luke9389. I not sure how to tackle this issue. In this solution, each worked is wrapped into a separate view, and then these are being flex wrapped in the next line. but if a single word would be very big it will wrap to the next line. So far I have no idea of how to break that word into child views.

**Comment by parasharrajat** [comment]
Updated @shawnborton. Let me know if that works.

**Comment by shawnborton** [comment]
@parasharrajat did you update the screenshots as well?

**Comment by parasharrajat** [comment]
Oh no. I will update them.

**Comment by parasharrajat** [comment]
@shawnborton I have updated the screens. Just I don't have IOS so I have put the image from [snap](https://snack.expo.io/@rajat_parashar/text-wrapping).

**Comment by shawnborton** [comment]
Hmm it looks like there is a larger gap between lines that have regular text + inline code, versus lines that are all inline code:
image

**Comment by parasharrajat** [comment]
@shawnborton I tried a lot of ways for this. **On IOS, the gap is equal** but on Android, the inline code is vertically top relative to the text so I have to push the inline code towards the bottom to Keep it centered which Increases the gap where inline code comes on the same line with the text.

After many trials, I found this solution to add border and border-radius to the inline code which has **this drawback on Android**. we(including other proposals) were only able to achieve the background-color for the inline code so far. I am open to suggestions if someone can look at the code and help me improve it.

**Comment by shawnborton** [comment]
Going to tag my pal @marcaaron for thoughts on this from a technical perspective, as I know we've gone through this one a few times now. 

It seems like we're getting close to the solution though. Is there any way you can make it so that both the normal text and inline code have a height of 20? If we can get the 20px tall containers to line up correctly, we hopefully can control the styles inside of them.

**Comment by parasharrajat** [comment]
I see that we set the line height of the whole text message then we can reach to almost equal gap but that text is not controlled by the inline code block and I don't think we don't want to change the line height for all messages. But let me see what I can do here.

**Comment by shawnborton** [comment]
Hmm I don't think I am quite following your comment. But I am pretty sure the line height for a default message is 20 right?

So my thinking is that if a block of text has a height/line height of 20, and then an inline code block is wrapped with a height of 20, those two elements should vertically align. Inside of the inline code block wrapper, we can make the code appear to be 18 tall with a margin top of 1 and a margin bottom of 1. Let me know if that makes sense.

**Comment by marcaaron** [comment]
Coming in a bit late to this PR and the overall direction and I'm curious... did we ever consider applying a solution like this in `react-native-render-html`?  IIRC we had a conversation going with a maintainer where they suggested a similar approach cc @AndrewGable 

As for styling advice, let's keep trying to achieve what we want for a bit. I don't have time to look into this today, but if we are still stuck down the line I'll jump in and see what I can do.

**Comment by parasharrajat** [comment]
@shawnborton So This is how it is rendered initially 
image

Now to make the text inline I have to move the boxes some px downwards which works fine for the box which has normal inline text such as line 2 but when we have a whole line of code block then the negative margin which we gave to pull it down is not necessary and thus as a side effect it unnecessarily pull the whole line downwards and reducing the gap between this and next line.

**Comment by shawnborton** [comment]
Hmm okay. Are all of your changes pushed to your branch? Maybe I can pull it down and test on my end and let you know what I find.

**Comment by parasharrajat** [comment]
@shawnborton That would be really great. yeah its pushed.

**Comment by shawnborton** [comment]
I tried playing around with this but I am stumped. I added some helper background colors to see where things were breaking down:
image


It seems like when the message is only inline code, we can make it behave correctly in terms of making everything 20px tall, but the inner code staying at 18px with the 1px gap around it within the 20px container. Notice the nice 2px gaps when this goes multi line:
image

When we compare a message that is text-only and is two lines versus a message that is inline-code-only and two lines, the size is pretty similar, though it looks like the inline code is shifted upwards of about 2px:
image

But the biggest thing that seems to be breaking is when we have a combination of text and inline code that flows to multiple lines:
image

Notice how there is a big gap between the two red blocks? Ideally that entire block should be 40px tall but it's actually 48px tall... so somehow there is a mystery 8px being added, and my eyes are telling me that it's an extra 4px being added to the bottom of each line. 

Anyways, I hope that is helpful in getting us closer, but I'm not entirely sure how to fix that mystery gap problem.

**Comment by parasharrajat** [comment]
@shawnborton @Luke9389. Breakthrough here. I have managed the gap. :partying_face: :partying_face:

**Comment by shawnborton** [comment]
Whoa, amazing Rajat!!

**Comment by shawnborton** [comment]
I think this is the behavior we want on all platforms, not just native. Is it possible to implement this everywhere?

**Comment by parasharrajat** [comment]
@shawnborton Yup. 
here is web/ Desktop
localhost_8080_r_68438375

**Comment by shawnborton** [comment]
Awesome, design-wise this looks great to me.

**Review by Luke9389 — CHANGES_REQUESTED** [review]
Hey Rajat, this looks awesome, nice job. 
I put some knit-picky whitespace reviews in, but otherwise this looks great!

**Comment by parasharrajat** [comment]
@Luke9389  Updated. Thanks

**Review by Luke9389 — APPROVED** [review]
Nailed it Rajat, way to go!

**Comment by isagoico** [comment]
## Chat - Sent code blocks are not properly centered

## Expected Result
Sent message and code blocks are vertically centered/alligned

## Actual Result
Code block is not centered vertically

## Action Performed
1. Open staging.expensify.cash and login
2. Open any conversation
3. Send any message within ` , to make them code blocks

## Platform 

iOS ✔️ 

**Build** 1.0.28-0

**Notes/Images/Video**

image

**Comment by roryabraham** [comment]
Since this is a deploy blocker, the fix for which is still in review, I'm going to revert this PR. @parasharrajat what this means for you is that you'll need to resubmit a new PR that's a combination of this PR (#2307) and the fix for the deploy blocker (#2527).

That way, we can hopefully move forward with a production deploy as soon as possible. Let me know if you have any questions.

**Comment by parasharrajat** [comment]
Sure

File: src/styles/styles.js [hunk 1]
```diff
@@ -1389,9 +1420,9 @@ const webViewStyles = {
             ...baseCodeTagStyles,
             paddingLeft: 5,
             paddingRight: 5,
-            paddingBottom: 2,
-            alignSelf: 'flex-start',
             fontFamily: fontFamily.MONOSPACE,
+            lineHeight: 18,
+            fontSize: 13,
```
Comment (s77rt) [inline src/styles/styles.js:1425]: The inline code block font size was being inherited. But after explicitly setting it to 13px, it's always rendered at 13px even if the parent is a `H1` where the font size is expected to be a little bigger. (Coming from https://github.com/Expensify/App/issues/30203)

**Pull request merged**